packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -528,9 +435,149 @@ contract ControllerHelper is UniswapControllerHelper, AaveControllerHelper, IERC
+            ControllerHelperDataType.CALLBACK_SOURCE(_callSource) ==
+            ControllerHelperDataType.CALLBACK_SOURCE.FLASHLOAN_REBALANCE_VAULT_NFT
+        ) {
+            // convert flashloaned WETH to ETH
+            IWETH9(weth).withdraw(_amount);
 
-            _checkPartialLpClose(data.vaultId, data.tokenId, data.liquidityPercentage);
+            (uint256 vaultId, ControllerHelperDataType.RebalanceVaultNftParams[] memory data) = abi.decode(
+                _calldata,
+                (uint256, ControllerHelperDataType.RebalanceVaultNftParams[])
+            );
+
+            // deposit collateral into vault and withdraw LP NFT
+            IController(controller).deposit{value: _amount}(vaultId);
+            IController(controller).withdrawUniPositionToken(vaultId);
+
+            console.log("vaultId", vaultId);
+
+            for (uint256 i; i < data.length; i++) {
+                if (
+                    data[i].rebalanceVaultNftType == ControllerHelperDataType.RebalanceVaultNftType.IncreaseLpLiquidity
+                ) {
+                    if (address(this).balance > 0) IWETH9(weth).deposit{value: address(this).balance}();
+
+                    // increase liquidity in LP position, this can mint wPowerPerp and increase
+                    ControllerHelperDataType.IncreaseLpLiquidityParam memory increaseLiquidityParam = abi.decode(
+                        data[i].data,
+                        (ControllerHelperDataType.IncreaseLpLiquidityParam)
+                    );
+
+                    ControllerHelperUtil.increaseLpLiquidity(
+                        controller,
+                        nonfungiblePositionManager,
+                        wPowerPerp,
+                        vaultId,
+                        increaseLiquidityParam,
+                        isWethToken0
+                    );
+                } else if (
+                    data[i].rebalanceVaultNftType == ControllerHelperDataType.RebalanceVaultNftType.DecreaseLpLiquidity
+                ) {
+                    // decrease liquidity in LP
+                    ControllerHelperDataType.DecreaseLpLiquidityParams memory decreaseLiquidityParam = abi.decode(
+                        data[i].data,
+                        (ControllerHelperDataType.DecreaseLpLiquidityParams)
+                    );
+
+                    ControllerHelperUtil.closeUniLp(
+                        nonfungiblePositionManager,
+                        ControllerHelperDataType.closeUniLpParams({
+                            tokenId: decreaseLiquidityParam.tokenId,
+                            liquidity: decreaseLiquidityParam.liquidity,
+                            liquidityPercentage: decreaseLiquidityParam.liquidityPercentage,
+                            amount0Min: decreaseLiquidityParam.amount0Min,
+                            amount1Min: decreaseLiquidityParam.amount1Min
+                        }),
+                        isWethToken0
+                    );
```

Comment:
Should the returned LP assets go back into vault here?

Reply:
What do you mean ?